cirq-core/cirq/circuits/moment.py

```diff
@@ -187,17 +192,23 @@ def with_operations(self, *contents: 'cirq.OP_TREE') -> 'cirq.Moment':
         """
         operations = list(self._operations)
         qubits = set(self._qubits)
+        measurement_key_objs = set(self._measurement_key_objs_())
+        control_keys = set(self._control_keys_())
         for op in op_tree.flatten_to_ops(contents):
             if any(q in qubits for q in op.qubits):
                 raise ValueError(f'Overlapping operations: {op}')
             operations.append(op)
             qubits.update(op.qubits)
+            measurement_key_objs |= protocols.measurement_key_objs(op)
+            control_keys |= protocols.control_keys(op)
 
         # Use private variables to facilitate a quick copy.
         m = Moment()
         m._operations = tuple(operations)
         m._qubits = frozenset(qubits)
         m._qubit_to_op = self._qubit_to_op.copy()
+        m._measurement_key_objs = measurement_key_objs
+        m._control_keys = frozenset(control_keys)
```

Comment:
Yeah this is broken, but I think it needs more discussion in a separate follow up.  The problem is that the protocol return value is `AbstractSet`, which means that you could return a set and mutate it and expect it to mutate the set on the object.  I'm not sure that was the right choice.  But even if it was, then Moment should definitely use `FrozenSet` since we don't want to make it mutable via these protocols.  But then restricting the return type from `AbstractSet` to `FrozenSet` is technically a "breaking" change. 

I've gone and used itertools.chain to hopefully only read once where possible across these methods.  Opened up #5298 for the frozen set issue.

Reply:
My understanding of `AbstractSet` is it's meant to indicate that the return value should not be mutated (`AbstractSet` itself does not include mutation methods), even though we might return a mutable subclass like `Set`. This relies on callers respecting the annotated type, even if the runtime type is mutable, which is not great but sort of par for the course in python (is anything ever _really_ immutable? :-) ). That said, if we annotate the type as `AbstractSet` we can still return `FrozenSet` if we're paranoid.

Reply:
Oh yeah `typing.Set` isthe type for the stdlib Set.  I'd forgot how butchered that namespace is.  

Then I think both should be frozensets so people don't shoot themselves in the foot mutating these things.

Reply:
OK, I updated to return frozenset for measurement_key_obj and control_keys.  Also moved to a pattern that utilized `union` on frozenset (just straight using `itertools.chain` isn't as good if the initial sets have values).  

Also confirmed these still give speedups.